### PR TITLE
Every operator unspills

### PR DESCRIPTION
**Pull request by ArturGajowy**
@pnowojski @findepi please review

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSourceFactory.java [hunk 1]
```diff
@@ -248,24 +247,24 @@ private void checkAllFuturesDone()
     }
 
     @Override
-    public CompletableFuture<LookupSource> readSpilledLookupSource(Session session, int partition)
+    public synchronized CompletableFuture<LookupSource> readSpilledLookupSource(Session session, int partition)
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSourceFactory.java]: The method as it stands now does only two things:

* `spilledLookupSources.get(partition)`
* `lookupSourceSpiller.getAllSpilledPages()`

the rest is a callback.

For me, it's not obvious what synchronization is needed for. Or -- maybe it is needed, but this fact is not related to the rest of the commit?

Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSourceFactory.java]: I see this is removed in next commit, so it's totally up to you what you do with my comment :)

File: presto-main/src/main/java/com/facebook/presto/operator/PartitionedConsumption.java [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.concurrent.CompletableFuture.allOf;
+
+public class PartitionedConsumption<T extends Closeable>
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/PartitionedConsumption.java]: I know this would probably violate checkstyle, but consider some short doc.

File: presto-main/src/main/java/com/facebook/presto/operator/PartitionedConsumption.java [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    {
+        return consumers;
+    }
+
+    public Partition<T> firstPartition()
+    {
+        return firstPartition;
+    }
+
+    public static class Partition<S extends Closeable>
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/PartitionedConsumption.java]: `final`?

File: presto-main/src/main/java/com/facebook/presto/operator/PartitionedConsumption.java [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+public class PartitionedConsumption<T extends Closeable>
+{
+    private final int consumers;
+    private final Function<Integer, CompletableFuture<T>> loader;
+    private final Partition<T> firstPartition;
+
+    PartitionedConsumption(int consumers, Collection<Integer> partitionNumbers, Function<Integer, CompletableFuture<T>> loader)
+    {
+        this.consumers = consumers;
+        this.loader = loader;
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/PartitionedConsumption.java]: `rNN` here and for `partitionNumbers` too

File: presto-main/src/main/java/com/facebook/presto/operator/PartitionedConsumption.java [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+        this.firstPartition = createPartitions(partitionNumbers);
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/PartitionedConsumption.java]: `firstPartition = createPartitions(..)` sounds like either the field or the method is not named accurately.

Would `createPartitionChain` be better?

File: presto-main/src/main/java/com/facebook/presto/operator/PartitionedConsumption.java [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+        firstPartition.previousReleased();
+        return firstPartition;
+    }
+
+    public int getConsumersCount()
+    {
+        return consumers;
+    }
+
+    public Partition<T> firstPartition()
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/PartitionedConsumption.java]: Consider naming it `getPartitionChain`, otherwise I will wonder were are the `secondPartition` and `thirdPartition` methods.

File: presto-main/src/main/java/com/facebook/presto/operator/PartitionedConsumption.java [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+        private final CompletableFuture<Void> requested;
+        private final CompletableFuture<Void> previousReleased;
+        private final CompletableFuture<S> loaded;
+        private final Optional<Partition<S>> next;
+
+        private int pendingReleases;
+
+        public Partition(
+                int consumersCount,
+                int number,
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/PartitionedConsumption.java]: `partitionNumber`

File: presto-main/src/main/java/com/facebook/presto/operator/PartitionedConsumption.java [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.concurrent.CompletableFuture.allOf;
+
+public class PartitionedConsumption<T extends Closeable>
+{
+    private final int consumers;
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/PartitionedConsumption.java]: `consumersCount` ?

File: presto-main/src/main/java/com/facebook/presto/operator/PartitionedConsumption.java [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+                int number,
+                Function<Integer, CompletableFuture<S>> loader,
+                Optional<Partition<S>> next)
+        {
+            this.number = number;
+            this.requested = new CompletableFuture<>();
+            this.previousReleased = new CompletableFuture<>();
+            this.loaded = allOf(requested, previousReleased)
+                    .thenCompose(ignored -> loader.apply(number));
+            this.next = next;
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/PartitionedConsumption.java]: `rNN`

File: presto-main/src/main/java/com/facebook/presto/operator/PartitionedConsumption.java [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+
+    public static class Partition<S extends Closeable>
+    {
+        private final int number;
+        private final CompletableFuture<Void> requested;
+        private final CompletableFuture<Void> previousReleased;
+        private final CompletableFuture<S> loaded;
+        private final Optional<Partition<S>> next;
+
+        private int pendingReleases;
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/PartitionedConsumption.java]: `@GuardedBy("this")`

File: presto-main/src/main/java/com/facebook/presto/operator/PartitionedConsumption.java [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+        }
+
+        public int number()
+        {
+            return number;
+        }
+
+        public CompletableFuture<S> load()
+        {
+            return markRequested().loaded;
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/PartitionedConsumption.java]: why not just
```
markRequested();
return loaded;
```

?

File: presto-main/src/main/java/com/facebook/presto/operator/PartitionedConsumption.java [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+        {
+            checkState(loaded.isDone());
+            pendingReleases--;
+            checkState(pendingReleases >= 0);
+            if (pendingReleases == 0) {
+                try {
+                    loaded.join().close();
+                }
+                catch (IOException e) {
+                    throw new RuntimeException("Error while releasing partition", e);
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/PartitionedConsumption.java]: Not sure what's the established practice here but
```
throw new RuntimeException("Error while releasing partition:" +e, e);
```

gives more useful exc msg

File: presto-main/src/main/java/com/facebook/presto/operator/PartitionedConsumption.java [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+                }
+                next().ifPresent(next -> next.previousReleased());
+            }
+        }
+
+        private boolean previousReleased()
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/PartitionedConsumption.java]: This method name should be a verb.

File: presto-main/src/main/java/com/facebook/presto/operator/SharedPartitionedMemoryContext.java [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+ */
+package com.facebook.presto.operator;
+
+import java.util.Optional;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static java.lang.String.format;
+
+public class SharedPartitionedMemoryContext
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/SharedPartitionedMemoryContext.java]: some doc?

File: presto-main/src/main/java/com/facebook/presto/operator/SharedPartitionedMemoryContext.java [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+{
+    private int reservations = 0;
+    private Optional<Integer> partitionKey;
+    private Optional<Long> inMemorySizeInBytes;
+    private Optional<TaskContext> taskContext;
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/SharedPartitionedMemoryContext.java]: `GuardedBy`

File: presto-main/src/main/java/com/facebook/presto/operator/SharedPartitionedMemoryContext.java [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+
+    public synchronized void reserve(int partitionKey, OperatorContext operatorContext, long inMemorySizeInBytes)
+    {
+        if (this.reservations == 0) {
+            this.partitionKey = Optional.of(partitionKey);
+            this.inMemorySizeInBytes = Optional.of(inMemorySizeInBytes);
+            TaskContext taskContext = operatorContext.getDriverContext().getPipelineContext().getTaskContext();
+            this.taskContext = Optional.of(taskContext);
+            operatorContext.reserveMemory(inMemorySizeInBytes);
+            operatorContext.transferMemoryToTaskContext(inMemorySizeInBytes);
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/SharedPartitionedMemoryContext.java]: Why not `taskContext.reserveMemory()` directly?

File: presto-main/src/main/java/com/facebook/presto/operator/SharedPartitionedMemoryContext.java [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+                "Inconsistent memory de-reservation. No current reservation present, " +
+                "requested free for %s bytes for partition %s.",
+                partitionKey,
+                inMemorySizeInBytes
+        ));
+        checkCurrentReservationPresent();
+        checkArgument(this.partitionKey.get() == partitionKey && this.inMemorySizeInBytes.get() == inMemorySizeInBytes,
+                "Inconsistent memory de-reservation. " + getInconsistenceDescription(partitionKey, inMemorySizeInBytes));
+        reservations--;
+        if (reservations == 0) {
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/SharedPartitionedMemoryContext.java]: The idea is, a suppose, as follows:

* `SpillledLookupJoiners` call `reserve()` and `reservations` goes up
* they call `free` and `reservations` goes down, the last one frees

What will happen if there are 2 `SpillledLookupJoiners` and
* first calls `reserve`
* first calls `free` before the second called `reserve`

?

Maybe this will work correctly (i.e. memory will be freed and re-reserved).. I dont know, but if so, this should be explcitly commented in the code, otherwise this looks like a race condition.

File: presto-main/src/main/java/com/facebook/presto/operator/SharedPartitionedMemoryContext.java [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+        }
+    }
+
+    public void checkCurrentReservationPresent()
+    {
+        checkState(this.partitionKey.isPresent() && this.inMemorySizeInBytes.isPresent(),
+                "partitionKey, inMemorySizeBytes and taskContext must be present whenever reservations > 0. That's an error.");
+    }
+
+    public String getInconsistenceDescription(int partitionKey, long inMemorySizeInBytes)
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/SharedPartitionedMemoryContext.java]: private

File: presto-main/src/main/java/com/facebook/presto/operator/SharedPartitionedMemoryContext.java [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+        reservations--;
+        if (reservations == 0) {
+            this.partitionKey = null;
+            this.inMemorySizeInBytes = null;
+            taskContext.get().freeMemory(inMemorySizeInBytes);
+            taskContext = null;
+        }
+    }
+
+    public void checkCurrentReservationPresent()
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/SharedPartitionedMemoryContext.java]: private

File: presto-main/src/main/java/com/facebook/presto/operator/SharedPartitionedMemoryContext.java [hunk 20]
```diff
@@ -0,0 +1,91 @@
...
+            this.inMemorySizeInBytes = Optional.of(inMemorySizeInBytes);
+            TaskContext taskContext = operatorContext.getDriverContext().getPipelineContext().getTaskContext();
+            this.taskContext = Optional.of(taskContext);
+            operatorContext.reserveMemory(inMemorySizeInBytes);
+            operatorContext.transferMemoryToTaskContext(inMemorySizeInBytes);
+        }
+        else {
+            checkCurrentReservationPresent();
+            checkArgument(this.partitionKey.get() == partitionKey && this.inMemorySizeInBytes.get() == inMemorySizeInBytes,
+                    "Inconsistent memory reservation. " + getInconsistenceDescription(partitionKey, inMemorySizeInBytes));
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/SharedPartitionedMemoryContext.java]: You could also check the second coming party has same `taskContext`.

File: presto-main/src/main/java/com/facebook/presto/spiller/BinaryFileSingleStreamSpiller.java [hunk 21]
```diff
@@ -105,6 +107,14 @@ private void checkNoSpillInProgress()
         return readPages();
     }
 
+    @Override
+    public CompletableFuture<List<Page>> getAllSpilledPages()
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/spiller/BinaryFileSingleStreamSpiller.java:121]: Maybe `getSpilledPagesAsync` or `getSpilledPagesFuture`?

File: presto-main/src/main/java/com/facebook/presto/operator/SharedPartitionedMemoryContext.java [hunk 22]
```diff
@@ -0,0 +1,91 @@
...
+    {
+        checkArgument(reservations > 0, format(
+                "Inconsistent memory de-reservation. No current reservation present, " +
+                "requested free for %s bytes for partition %s.",
+                partitionKey,
+                inMemorySizeInBytes
+        ));
+        checkCurrentReservationPresent();
+        checkArgument(this.partitionKey.get() == partitionKey && this.inMemorySizeInBytes.get() == inMemorySizeInBytes,
+                "Inconsistent memory de-reservation. " + getInconsistenceDescription(partitionKey, inMemorySizeInBytes));
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/operator/SharedPartitionedMemoryContext.java]: "Inconsistent memory size on free"

File: presto-main/src/main/java/com/facebook/presto/operator/SharedPartitionedMemoryContext.java (hunk 14 above)
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/operator/SharedPartitionedMemoryContext.java]: In what sense is this context "partitioned".

File: presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java [hunk 23]
```diff
@@ -110,33 +112,29 @@ public void finish()
...
-                spiller = Optional.empty();
-            }
-            else {
-                // last LookupJoinOperator might not spilled at all, but other parallel joins could
-                // so we have to load spiller if this operator hasn't used one yet
-                ensureSpillerLoaded();
-            }
+            lookupPartition = Optional.of(
+                    lookupSourceFactory.beginLookupSourceUnspilling(lookupJoinsCount.get(), operatorContext.getSession()));
+            ensureSpillerLoaded();
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java]: Where is return value of `ensureSpillerLoaded` used?

File: presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java [hunk 24]
```diff
@@ -110,33 +112,29 @@ public void finish()
...
     {
         checkState(spiller.isPresent());
-        checkState(spilledPartitions.isPresent());
-        if (spilledPartitions.get().hasNext()) {
-            int currentSpilledPartition = spilledPartitions.get().next();
+        checkState(lookupPartition.isPresent());
 
+        spilledLookupJoiner.ifPresent(joiner -> {
+            joiner.finish(sharedPartitionedMemoryContext);
+            lookupPartition = lookupPartition.get().next();
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java]: Relation between this callback body and `spilledLookupJoiner` is less than obvious. I understand, the presence of `spilledLookupJoiner` is used as a control variable.

* using plain `if` would be more readable
* a comment explaining the control flow would help me understand

File: presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSourceFactory.java [hunk 25]
```diff
@@ -247,24 +252,43 @@ private void checkAllFuturesDone()
     }
 
     @Override
-    public synchronized CompletableFuture<LookupSource> readSpilledLookupSource(Session session, int partition)
+    public synchronized Partition<LookupSource> beginLookupSourceUnspilling(int consumersCount, Session session)
     {
-        SingleStreamSpiller lookupSourceSpiller = spilledLookupSources.get(partition);
+        checkState(hasSpilled());
+        if (lookupSourceUnspilling.isPresent()) {
+            checkState(lookupSourceUnspilling.get().getConsumersCount() == consumersCount, "All consumers must pass the same consumersCount");
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSourceFactory.java]: 👍

File: presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSourceFactory.java [hunk 26]
```diff
@@ -247,24 +252,43 @@ private void checkAllFuturesDone()
...
+        }
+        else {
+            Set<Integer> partitionNumbers = spilledLookupSources.keySet();
+            Function<Integer, CompletableFuture<LookupSource>> partitionLoader = partitionNumber -> {
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSourceFactory.java]: what about extracting to a method? Or instead of `readSpilledLookupSource` ?

File: presto-main/src/main/java/com/facebook/presto/operator/PartitionedConsumption.java [hunk 27]
```diff
@@ -0,0 +1,133 @@
...
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.concurrent.CompletableFuture.allOf;
+
+public class PartitionedConsumption<T extends Closeable>
+{
+    private final int consumers;
+    private final Function<Integer, CompletableFuture<T>> loader;
+    private final Partition<T> firstPartition;
+
+    PartitionedConsumption(int consumers, Collection<Integer> partitionNumbers, Function<Integer, CompletableFuture<T>> loader)
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/operator/PartitionedConsumption.java]: Can we really know the consumers count upfront? Are all operators alway started?

Comment (losipiuk) [reply presto-main/src/main/java/com/facebook/presto/operator/PartitionedConsumption.java]: I would change consumers to consumersCount

File: presto-main/src/main/java/com/facebook/presto/spiller/BinaryFileSingleStreamSpiller.java [hunk 28]
```diff
@@ -105,6 +107,14 @@ private void checkNoSpillInProgress()
...
+    {
+        return MoreFutures.toCompletableFuture(executor.submit(() ->
+                ImmutableList.copyOf(getSpilledPages())
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/spiller/BinaryFileSingleStreamSpiller.java]: Aren't you overly un-eager here? `getSpilledPages()` reads `Page`-s lazily.

File: presto-main/src/main/java/com/facebook/presto/operator/PartitionedConsumption.java (hunk 3 above)
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/operator/PartitionedConsumption.java]: I would split this class to two.

PartitionChain (as suggested by @findepi) with methods:
 * CompletableFuture<Optional<Partition>> loadNextPartition()

And Partition with method:
 * release()

**Review by losipiuk — COMMENTED** [review]
partial review

File: presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java [hunk 29]
```diff
@@ -110,33 +112,29 @@ public void finish()
         lookupJoiner.finish();
 
         if (!finishing && lookupSourceFactory.hasSpilled()) {
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java]: Do we have a race here?

* one operator gets `finish()`
* second operator starts spilling

the ordering of these two actions has huge consequences (if they happen as I put them, the first operator will be counted among expected consumer counts, but will never show up).

Maybe such an ordering is not possible, but this should be asserted in the code. For example:

```
if (!finishing) {
 hasSpilled = lookupSourceFactory.tellTheFactoryIFinishedAndAskIfAnyoneHasSpilledOhAndTellItToPreventAnyFutureSpillsBtw(); 
 if (hasSpilled) {
    ....
```

File: presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java [hunk 30]
```diff
@@ -236,7 +231,7 @@ public Page getOutput()
                 return null;
             }
             SpillledLookupJoiner joiner = spilledLookupJoiner.get();
-            operatorContext.setMemoryReservation(joiner.getInMemorySizeInBytes());
+            joiner.reserveMemory(sharedPartitionedMemoryContext, operatorContext);
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java]: Can `spilledLookupJoiner.get().isFinished()` at this point?

File: presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperatorFactory.java [hunk 31]
```diff
@@ -123,6 +125,7 @@ private LookupJoinOperatorFactory(LookupJoinOperatorFactory other)
         referenceCount = other.referenceCount;
         outerOperatorFactory = other.outerOperatorFactory;
         probeHashGenerator = other.probeHashGenerator;
+        sharedPartitionedMemoryContext = other.sharedPartitionedMemoryContext;
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperatorFactory.java]: Not sure what are the semantics of `LookupJoinOperatorFactory.duplicate`, but are you sure passing *same* `sharedPartitionedMemoryContext` should be here?

I looked into usages but I didn't raise confidence about this.

File: presto-main/src/main/java/com/facebook/presto/operator/PartitionedConsumption.java [hunk 32]
```diff
@@ -0,0 +1,133 @@
...
+        }
+
+        public synchronized void release()
+        {
+            checkState(loaded.isDone());
+            pendingReleases--;
+            checkState(pendingReleases >= 0);
+            if (pendingReleases == 0) {
+                try {
+                    loaded.join().close();
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/PartitionedConsumption.java]: `getFutureValue(loaded).close()`

File: presto-main/src/main/java/com/facebook/presto/spiller/BinaryFileSingleStreamSpiller.java (hunk 21 above)
Comment (ArturGajowy) [reply presto-main/src/main/java/com/facebook/presto/spiller/BinaryFileSingleStreamSpiller.java:121]: I proposed the latter but @pnowojski resisted :)

File: presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java (hunk 23 above)
Comment (ArturGajowy) [reply presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java]: elsewhere :D

File: presto-main/src/main/java/com/facebook/presto/operator/SharedPartitionedMemoryContext.java (hunk 16 above)
Comment (ArturGajowy) [reply presto-main/src/main/java/com/facebook/presto/operator/SharedPartitionedMemoryContext.java]: Dunno, just followed @pnowojski's advice here

File: presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java (hunk 24 above)
Comment (ArturGajowy) [reply presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java]: I frowned at this part of code as well when writing it, but didn't want to overdo it. I think the PartitionChain concept that emerged in review (and I considered during writing the code) will help here

File: presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperatorFactory.java (hunk 31 above)
Comment (ArturGajowy) [reply presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperatorFactory.java]: I think it should be duplicated, but @pnowojski should have a look as well. The relevant line is `LocalExecutionPlanner:435`

File: presto-main/src/main/java/com/facebook/presto/operator/SharedPartitionedMemoryContext.java (hunk 14 above)
Comment (ArturGajowy) [reply presto-main/src/main/java/com/facebook/presto/operator/SharedPartitionedMemoryContext.java]: I renamed the class to `SharedMemoryContext` and the `partitionKey` paremeter to `reservationKey`.

File: presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java (hunk 29 above)
Comment (ArturGajowy) [reply presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java]: added a comment to explain why this is not possible ATM

File: presto-main/src/main/java/com/facebook/presto/operator/PartitionedConsumption.java (hunk 3 above)
Comment (ArturGajowy) [reply presto-main/src/main/java/com/facebook/presto/operator/PartitionedConsumption.java]: Made PartitionedConsumption have a `getPartitions() : Iterable<Partition<T>>`

File: presto-main/src/main/java/com/facebook/presto/operator/SharedPartitionedMemoryContext.java (hunk 17 above)
Comment (ArturGajowy) [reply presto-main/src/main/java/com/facebook/presto/operator/SharedPartitionedMemoryContext.java]: It's not possible ATM due to PartitionedConsumption coordinating the partions consumption :D Also, if this were to happen - we'd just have a little bit less effective sanity check.

File: presto-main/src/main/java/com/facebook/presto/spiller/BinaryFileSingleStreamSpiller.java (hunk 28 above)
Comment (ArturGajowy) [reply presto-main/src/main/java/com/facebook/presto/spiller/BinaryFileSingleStreamSpiller.java]: added `Disallow spilling after first read in BinaryFileSingleStreamSpiller`

**Comment by ArturGajowy** [comment]
@losipiuk @findepi comments applied please have a look again

File: presto-main/src/main/java/com/facebook/presto/spiller/BinaryFileSingleStreamSpiller.java (hunk 21 above)
Comment (pnowojski) [reply presto-main/src/main/java/com/facebook/presto/spiller/BinaryFileSingleStreamSpiller.java:121]: There is no and there shouldn't be synchronous method for this, so why differentiate it in method name, when signature does the same trick?

File: presto-main/src/main/java/com/facebook/presto/spiller/BinaryFileSingleStreamSpiller.java (hunk 28 above)
Comment (pnowojski) [reply presto-main/src/main/java/com/facebook/presto/spiller/BinaryFileSingleStreamSpiller.java]: I don't get this answer :) Code seems reasonable, since it's submitted to the executor, but I don't understand  this conversation and this `Disallow spilling after first read in BinaryFileSingleStreamSpiller` commit.

File: presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java [hunk 33]
```diff
@@ -190,9 +190,9 @@ public void addInput(Page page)
 
     private Page spillAndMaskSpilledPositions(Page page)
     {
-        ensureSpillerLoaded();
+        PartitioningSpiller spiller = ensureSpillerLoaded();
```
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java]: You are hiding instance variable, maybe rename spiller -> partitioningSpiller?

File: presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java [hunk 34]
```diff
@@ -202,7 +202,7 @@ private Page spillAndMaskSpilledPositions(Page page)
         return mapPage(unspilledPositions, page);
     }
 
-    private void ensureSpillerLoaded()
+    private PartitioningSpiller ensureSpillerLoaded()
```
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java]: I don't like that this method returns something, while it operates on state. Signature suggests that returned value is a local variable/value, while in the reality it is stored on state. 

That's why I made it `void` and it was on purpose. If you really have so strong feelings about this change then whatever, but when I touch this code next time, I should revert this change and we would be stuck in a endless loop of reverting? :)

edit: especially that there are two places which call this method and in second one you ignore returned value. So I really don't see how this change simplify anything. IMO it only adds a confusion.

File: presto-main/src/main/java/com/facebook/presto/spiller/BinaryFileSingleStreamSpiller.java [hunk 35]
```diff
@@ -90,6 +92,7 @@ public BinaryFileSingleStreamSpiller(
     public CompletableFuture<?> spill(Iterator<Page> pageIterator)
     {
         checkNoSpillInProgress();
+        checkNotFrozen();
```
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/spiller/BinaryFileSingleStreamSpiller.java]: `checkIsWritable`?

File: presto-main/src/main/java/com/facebook/presto/spiller/BinaryFileSingleStreamSpiller.java [hunk 36]
```diff
@@ -100,16 +103,24 @@ private void checkNoSpillInProgress()
...
+    private void checkNotFrozen()
+    {
+        checkState(!frozen.get(), "Spilling no longer allowed. " +
+                "The spiller has been frozen on first read for subsequent reads to be consistent");
+    }
+
     @Override
     public Iterator<Page> getSpilledPages()
     {
+        frozen.set(true);
```
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/spiller/BinaryFileSingleStreamSpiller.java]: `writable.set(false)`?
